File: derivatives/netnode/store.py
"""Crash-safe on-disk block store for the X-chain nodes (Path B, Stage 1).

Every valid block a node accepts is appended, length-prefixed, to `<datadir>/blocks.dat`,
flushed + fsync'd. On startup the records are replayed into a `chainsync.Chain` (genesis
first, then `process_block` for the rest — whose orphan handling tolerates out-of-order
records). A truncated trailing record (from a crash mid-write) is ignored. This gives a node
**restart recovery**: it reloads its chain from disk instead of re-syncing from scratch.
Evidence: MODEL.
"""

from __future__ import annotations

import os
import pathlib
import struct


class BlockStore:
    def __init__(self, datadir: str | pathlib.Path):
        self.dir = pathlib.Path(datadir)
        self.dir.mkdir(parents=True, exist_ok=True)
        self.path = self.dir / "blocks.dat"
        self._f = open(self.path, "ab", buffering=0)     # append, unbuffered

    def append(self, raw: bytes) -> None:
        self._f.write(struct.pack("<I", len(raw)) + raw)
        self._f.flush()
        os.fsync(self._f.fileno())

    def read_all(self) -> list[bytes]:
        if not self.path.exists():
            return []
        data = self.path.read_bytes()
        out, i = [], 0
        while i + 4 <= len(data):
            (n,) = struct.unpack_from("<I", data, i)
            i += 4
            if i + n > len(data):                        # truncated tail from a crash -> stop
                break
            out.append(data[i:i + n])
            i += n
        return out

    def count(self) -> int:
        return len(self.read_all())

    def close(self) -> None:
        try:
            self._f.close()
        except OSError:
            pass

File: derivatives/validator-rs/src/store.rs
//! A crash-safe on-disk block store — mirrors `netnode/store.py`. NOT money.
//!
//! Length-prefixed append (`[len:4 LE][raw]`), fsync'd; `read_all` ignores a crash-truncated tail
//! (a partial final record), so a node recovers cleanly after an ill-timed crash.

use std::fs::{File, OpenOptions};
use std::io::{self, Read, Write};
use std::path::{Path, PathBuf};

pub struct BlockStore {
    path: PathBuf,
    file: File,
}

impl BlockStore {
    pub fn open(dir: &Path) -> io::Result<Self> {
        std::fs::create_dir_all(dir)?;
        let path = dir.join("blocks.dat");
        let file = OpenOptions::new().create(true).read(true).append(true).open(&path)?;
        Ok(BlockStore { path, file })
    }

    /// Append a block and flush it to disk.
    pub fn append(&mut self, raw: &[u8]) -> io::Result<()> {
        self.file.write_all(&(raw.len() as u32).to_le_bytes())?;
        self.file.write_all(raw)?;
        self.file.sync_all()?;
        Ok(())
    }

    /// All whole records, ignoring a crash-truncated final one.
    pub fn read_all(&self) -> io::Result<Vec<Vec<u8>>> {
        let mut data = Vec::new();
        File::open(&self.path)?.read_to_end(&mut data)?;
        let mut out = Vec::new();
        let mut i = 0;
        while i + 4 <= data.len() {
            let n = u32::from_le_bytes(data[i..i + 4].try_into().unwrap()) as usize;
            i += 4;
            if i + n > data.len() {
                break; // truncated tail -> ignore
            }
            out.push(data[i..i + n].to_vec());
            i += n;
        }
        Ok(out)
    }
}
